### PR TITLE
Making coordinate variables searchable 

### PR DESCRIPTION
**Pull request by charles-turner-1**
## Functionality changes in this branch:
- `source.builders.BaseBuilder.parse_access_ncfile` indexes all variables (`for var in ds.data_vars` => `for var in ds.variables`), rather than just data variables, & tests have been updated to reflect additional data obtained.
    - NOTE: This is dependent on changes to [intake-esm](https://github.com/charles-turner-1/intake-esm/tree/issue_660) proposed by @dougiesquire in [isssue 660](https://github.com/intake/intake-esm/issues/660).
    - Coordinate variables are accessible through the variable flag in the access-nri intake catalogue, rather than a separate coord flag, as suggested in #201 by @anton-seaice.
        - When searching for coordinate variables only, only coordinate variables are returned. No concatenation of assets is performed - only the first asset in the dataset is. This avoids crashing resulting from `xr.combine_by_coords`.
        - When searching for coordinate & data_variables, the data & coordinate variables searched, as well as coordinate variables that the searched data variables depend on.
        - Other searches will return all data & coordinate variables, as before. 
   
## Refactoring
- Updated `source.builders.BaseBuilder` to use a dataclass (`source.utils._AccessNCFileInfo`) to pass around information about variables, rather than an outputs tuple.
- Added a `_DataVarInfo` dataclass to hold attribute information from data variables, rather than appending them all into separate lists passed to the outputs tuple.

## Minor changes
- Add a bunch of type checking (type annotations, added mypy to pre-commit, etc)
- Replaced clauses of the pattern
    ```python
    try:
        x = x[key]
    except:
       pass
    ``` 
   with
    ```
    x = x.get(key,x)
    ```
- A couple of minor docstring updates

### intake-esm
The code changes in this branch should lead to catalogues being built with coordinate variables included in their variable flag: eg
```python 

from access_nri_intake.source.builders import AccessOm2Builder

builder = AccessOm2Builder(
    path="/g/data/ik11/outputs/access-om2-01/01deg_jra55v13_ryf9091"
)

builder.build()

CATALOG_PATH = ...

builder.save(
    directory='CATALOG_PATH',
    name="ACCESS-OM2_01deg_jra55v13_ryf9091", 
    description="...",
)
```
will build a catalog which coordinate variables are searchable from. However, unless you're working with a version of intake-esm which also has coordinate variables searchable (alterations noted in Functionality changes), this will present two issues:

1. Files comprising *only* coordinate variables, eg. `ocean_grid.nc` files, will not be indexed.
2. Searchs purely for a coordinate variable will fail when calling `.to_dask()` will fail on coordinate indexing if a multi file dataset is obtained from the search, or return an empty dataset if a single file is returned. 

Issue 1. appears to be [ESMDataSource._open_dataset](https://github.com/intake/intake-esm/blob/6ba67e1df72ebbc80b6cd5f78fd73ddb2cee564a/intake_esm/source.py#L219) thinking the .nc files are empty unless they contain coordinate variables & ignoring them.
Issue 2. is the result of the [requested_variables](https://github.com/intake/intake-esm/blob/6ba67e1df72ebbc80b6cd5f78fd73ddb2cee564a/intake_esm/source.py#L84) used in intake-esm. 

Neither of these is testable or fixable in access-nri-intake-catalog, as these issues are handled by intake-esm.

### PR size
This PR contains a couple of code changes, as well as some other stuff (typing/pre-commit, etc). I'm happy to split this apart into separate pull requests if that makes things more tractable to review.

### CODE REVIEW

**Comment by charles-turner-1** [comment]
I've been through and linked a bunch of issues that I think the changes here resolve - about to verify.

**Review by marc-white — COMMENTED** [review]
Correct me if I'm wrong, but this PR doesn't actually do anything to read the coordinate variables itself (compared to the read in the `main` version of the code currently does) - it's simply been prepped to handle the coordinates once `intake-esm` gets updated?

File: src/access_nri_intake/source/builders.py [hunk 1]
```diff
@@ -289,51 +314,31 @@ def parse_access_ncfile(cls, file, time_dim="time"):
...
         ) as ds:
-            variable_list = []
-            variable_long_name_list = []
-            variable_standard_name_list = []
-            variable_cell_methods_list = []
-            variable_units_list = []
-            for var in ds.data_vars:
+            dvars = _DataVarInfo()
+
+            for var in ds.variables:
```
Comment (charles-turner-1) [inline src/access_nri_intake/source/builders.py:319]: New behaviour - scans all variables in dataset, including coordinate variables

File: src/access_nri_intake/source/builders.py [hunk 2]
```diff
@@ -289,51 +314,31 @@ def parse_access_ncfile(cls, file, time_dim="time"):
             decode_times=False,
             decode_coords=False,
         ) as ds:
-            variable_list = []
-            variable_long_name_list = []
-            variable_standard_name_list = []
-            variable_cell_methods_list = []
-            variable_units_list = []
-            for var in ds.data_vars:
```
Comment (charles-turner-1) [inline src/access_nri_intake/source/builders.py:297]: Previous behaviour - scans all data_vars in dataset

**Comment by charles-turner-1** [comment]
Not precisely - it does additionally scan .nc files for coordinates (see above), but crucially also requires the esm_datastore object to do the same. If we have the same behaviour in the esm_datastore (defined in intake-esm) there this PR won't affect the variables we read.

**Comment by charles-turner-1** [comment]
I've checked, and the changes in this PR don't resolve #117 or #112.

In both instances, the issues specifically result from attempted concatenation of **grid** variables, rather than coordinate variables. Grid variables are stored as data variables in the netCDF files that hold them, but are don't depend on time & so the time variable is dropped before concatenation, which is the source of the 'no dimension to concatenate along' errors. 

I've changed this PR & the concomitant PR on intake-esm to draft status whilst I work out how closely related these issues are & whether we can neatly tie together all the changes we need.

File: src/access_nri_intake/source/utils.py [hunk 3]
```diff
@@ -4,16 +4,96 @@
...
+    standard_name_list: list[str] = field(default_factory=list)
+    cell_methods_list: list[str] = field(default_factory=list)
+    units_list: list[str] = field(default_factory=list)
+
+    def append_attrs(self, var: str, attrs: dict) -> None:
+        """
+        Append attributes to the DataVarInfo object, if the attribute has a
+        'long_name' key.
+
+        TODO: Why do we need a long name key? seems important
```
Comment (marc-white) [inline src/access_nri_intake/source/utils.py:71]: Is the code using the presence of 'long_name' as a proxy for "this is a real variable, not just additional info"?

**Comment by charles-turner-1** [comment]
I've split this into a series of stacked PR's: 
#221: Adds typing, mypy type check to pre-commit. Merge into main, addresses #13.
#222: Refactors `BaseBuilder.parse_access_ncfile` to return a dataclass, rather than a tuple. Closes #186 and the PR merges into the branch introduced by #221 
#212: (This PR) - changes `data_vars` => `variables` and updates the relevant tests to ensure we are catching all the correct coordinate variables. PR merges into the branch introduced by #222.

**Review by marc-white — APPROVED** [review]
Approved as ready, although are we in a position to merge before Intake-ESM is updated?

**Comment by charles-turner-1** [comment]
I'll go back and triple check before I merge, but it should be the case that the catalogue will build exactly as before if we don't update intake-esm.

**Comment by charles-turner-1** [comment]
The concomitant changes have been merged into intake-esm ([link](https://github.com/intake/intake-esm/pull/681#event-14968541423)), so we should be close to ready to go with making this the default behaviour.

**Comment by rbeucher** [comment]
Great 👍

**Pull request merged**